[WIP] Flamelet approach for premixed flames

bigfooted opened the pull request:
## Proposed Changes
*Give a brief overview of your contribution here in a few sentences.*
 This PR introduces a tabulated chemistry approach. The focus for this PR is on laminar and premixed flames in the incompressible framework, but it is generalizable to non-premixed and partially premixed flames, turbulent flames and compressible flames. 
In this approach, 2 transport equations are solved for progress variable and total enthalpy. The lookup table retrieves quantities (temperature, density, viscosity, diffusivity, source terms, etc) as a function of these independent control variables. 2 additional independent transport equations are solved for the species NO and CO.

## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*

## Nijso Checklist
- [ ] Replace bloated initialization (burnt and unburnt region) with python wrapper
- [ ] Check Conjugate Heat Transfer
- [ ] Check adjoint optimization workflow
- [x] Check axisymmetric setup
- [x] Check diffusivity
- [x] Check primal restart
- [x] Check adjoint restart
- [x] Adjoint comparison with FD

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [ ] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

pcarruscag commented:
Sorry to be a pain but can you keep the PR's separate? If you need a combined branch for testing create a third branch for the time being.

bigfooted commented:
Agreed, it's already big as it is, so merging this PR with the 3D table is going to be a mess. Besides we need to add a ton of tests still

bigfooted commented:
@EvertBunschoten can you revert this asap, because now nothing works anymore. I don't like what you did here.

bigfooted commented:
@EvertBunschoten, all: My proposal is to first put this PR in develop, then the PR for the 3D lookup table. The multilayer perceptron PR is independent, so that should be fine.

EvertBunschoten commented:
> 

@bigfooted Apologies for trying to do it all at once. We can indeed make it work with the 2D tables first and later implement the preferential diffusion physics.

pcarruscag commented:
@bigfooted @EvertBunschoten Well fwiw the simplest way is to checkout a new branch from before the merge and open a new PR, there are no coments here yet, so its fine.
(If you git revert the merge it will be a pain to then merge the other PRs)

bigfooted commented:
> @bigfooted @EvertBunschoten Well fwiw the simplest way is to checkout a new branch from before the merge and open a new PR, there are no coments here yet, so its fine. (If you git revert the merge it will be a pain to then merge the other PRs)

I did a git reset --merge <commit-id> to go back to my latest commit. I think this completely removed Evert's merge without any residual effects. So should we create a new PR or not? Your 'its fine' comment has ambiguous meaning :-)

pcarruscag commented:
If it doesnt show on the commit history it should be fine

Cristopher-Morales commented on SU2_CFD/src/solvers/CIncEulerSolver.cpp:
```diff
@@ -330,6 +331,15 @@ void CIncEulerSolver::SetNondimensionalization(CConfig *config, unsigned short i
 
+    case FLUID_FLAMELET:
+
+      config->SetGas_Constant(UNIVERSAL_GAS_CONSTANT / (config->GetMolecular_Weight() / 1000.0));
+      Pressure_Thermodynamic = Density_FreeStream * Temperature_FreeStream * config->GetGas_Constant();
```
From the Fluid-Mixture model , I have experienced the thermodynamic pressure influences the density at the inlet boundaries. So it would be better to provide the pressure_thermodynamic as a user value or a default value as 101325 [pa]. and based on this value compute the density.

bigfooted commented:
@EvertBunschoten  Thank you for not leaving us with all these alarming red crosses in the regression tests too long :-)
This week I will work on a small pywrapper addition to deal with initial conditions, then we can get rid of the flame-specific initialization. That will reduce the PR by a couple of hundred lines and even add functionality. But I'll create a separate PR for this.

bigfooted commented:
@EvertBunschoten @danielmayer @Cristopher-Morales I have now modified the flamelet approach such that it uses the diffusion coefficient D from the LUT file. In this way, the variable mass_diffusivity contains the same variable/same units for the flamelet approach as for the species transport approach. Note that this means you now need diffusion coefficient "D" in the LUT instead of "rho*D".
@EvertBunschoten I have removed the hard-coded computation of diffusivity from the unity Lewis approach.

bigfooted commented:
@danielmayer @EvertBunschoten @Cristopher-Morales I ran the scripts that we have for the species transport validation (from @TobiKattmann). Primal restart, adjoint restart and the comparison between adjoint and finite difference gradients all seem to work fine for the current PR. I will add some additional regression tests this week to make sure we do not break anything while polishing the code (specifically adding an axisymmetric case and a CHT case).

bigfooted commented:
@EvertBunschoten : no hard coded unity Lewis number approach

danielmayer commented on SU2_CFD/src/fluid/CFluidFlamelet.cpp:
```diff
@@ -72,19 +72,16 @@ CFluidFlamelet::CFluidFlamelet(CConfig* config, su2double value_pressure_operati
-    table_source_names[1 + n_user_scalars + i_aux] = config->GetUserSourceName(i_aux + 1);
+  for(size_t i_aux = 0; i_aux < n_user_scalars; i_aux++) {
+    /*--- Order of the source terms: S_prod_1, S_cons_1, S_prod_2, S_cons_2, ...---*/
+    table_source_names[1 + i_aux]                  = config->GetUserSourceName(i_aux*n_user_scalars);
```
Good catch but I think that should read:

`config->GetUserSourceName(i_aux*2);` and
`config->GetUserSourceName(i_aux*2 + 1);`

EvertBunschoten replied:
Good point!

bigfooted commented:
@EvertBunschoten please stop committing to this pull request. From now on, I will automatically remove all your commits without looking at them.

bigfooted commented:
discadj_arina2k fails with some mpi error, but it does not fail when I run it on my local machine. Not sure what is going on here.
There is another case that sometimes fails due to the tight tolerance.

danielmayer commented on SU2_CFD/src/fluid/CFluidFlamelet.cpp:
```diff
@@ -72,7 +72,7 @@ CFluidFlamelet::CFluidFlamelet(CConfig* config, su2double value_pressure_operati
   table_source_names.resize(n_table_sources);
   table_sources.resize(n_table_sources);
-  table_source_names[I_SRC_TOT_PROGVAR] = "ProdRateTot_PV";
+  table_source_names[I_SRC_TOT_PROGVAR] = "ProdRateTot-PV";
```
I would like to call it ProdRateTot_PV in the latest dragon format. The logic is <quantity>_<specifier> and inside <specifier> we can use "-" to separate words. Also, e.g.:  ProdRatePos_Y-CO. I know this is an inconvenience
for existing dragon tables with the other nomenclature. My proposal is to rename this in your existing tables now for the last time and in the future we will handle changes to the hard coded variable names through version numbers of the dragon file. Hope that works for everybody. Please let me know.

bigfooted commented:
I will close this pull request every time Evert commits something.
